chain/epoch-manager/src/tests/mod.rs

```diff
@@ -2928,6 +2928,183 @@ fn test_max_kickout_stake_ratio() {
+    );
+    assert_eq!(kickouts, HashMap::new());
+}
+
+#[test]
+/// Tests the case where a chunk validator has low endorsement stats and is kicked out (not exempted).
+/// In this test, first 3 accounts are block and chunk producers and next 2 are chunk validator only.  
+fn test_chunk_validator_kicked_out_for_low_endorsement() {
+    if !ProtocolFeature::ChunkEndorsementsInBlockHeader.enabled(PROTOCOL_VERSION) {
+        return;
+    }
+    let mut epoch_config = epoch_config(5, 2, 4, 80, 80, 80).for_protocol_version(PROTOCOL_VERSION);
+    let accounts = vec![
+        ("test0".parse().unwrap(), 1000),
+        ("test1".parse().unwrap(), 1000),
+        ("test2".parse().unwrap(), 1000),
+        ("test3".parse().unwrap(), 1000),
+        ("test4".parse().unwrap(), 1000),
+        ("test5".parse().unwrap(), 1000),
+    ];
+    let epoch_info = epoch_info(0, accounts, vec![0, 1, 2, 3], vec![vec![0, 1], vec![0, 2]]);
+    let block_stats = HashMap::from([
+        (0, ValidatorStats { produced: 90, expected: 100 }),
+        (1, ValidatorStats { produced: 90, expected: 100 }),
+        (2, ValidatorStats { produced: 90, expected: 100 }),
+        (3, ValidatorStats { produced: 0, expected: 0 }),
+    ]);
+    let chunk_stats = HashMap::from([
+        (
+            0,
+            HashMap::from([
+                (0, ChunkStats::new_with_production(90, 100)),
+                (1, ChunkStats::new_with_production(90, 100)),
+                (3, ChunkStats::new_with_endorsement(0, 0)),
+                (4, ChunkStats::new_with_endorsement(10, 100)),
+                (5, ChunkStats::new_with_endorsement(90, 100)),
+            ]),
+        ),
+        (
+            1,
+            HashMap::from([
+                (0, ChunkStats::new_with_production(90, 100)),
+                (2, ChunkStats::new_with_production(90, 100)),
+                (3, ChunkStats::new_with_endorsement(0, 0)),
+                (4, ChunkStats::new_with_endorsement(10, 100)),
+                (5, ChunkStats::new_with_endorsement(90, 100)),
+            ]),
+        ),
+    ]);
+
+    let prev_validator_kickout =
+        HashMap::from([("test3".parse().unwrap(), ValidatorKickoutReason::Unstaked)]);
+    // At most 40% of total stake can be kicked out
+    epoch_config.validator_max_kickout_stake_perc = 40;
+    let (_, kickouts) = EpochManager::compute_validators_to_reward_and_kickout(
+        &epoch_config,
+        &epoch_info,
+        &block_stats,
+        &chunk_stats,
+        &HashMap::new(),
+        &prev_validator_kickout,
+    );
+    assert_eq!(
+        kickouts,
+        HashMap::from([(
+            "test4".parse().unwrap(),
+            NotEnoughChunkEndorsements { produced: 20, expected: 200 }
+        ),])
+    );
+}
+
+#[test]
+/// Tests that a validator is not kicked out due to low endorsement only (as long as it produces most of its blocks and chunks).
+fn test_block_and_chunk_producer_not_kicked_out_for_low_endorsements() {
```

Comment:
We currently have
```
let chunk_validator_only =
                stats.block_stats.expected == 0 && stats.chunk_stats.expected() == 0;
if chunk_validator_only
                && stats
                    .chunk_stats
                    .endorsement_stats()
                    .less_than(chunk_validator_only_kickout_threshold)
            {
...
```

so the endorsements kickout is not currently applied to BP and CP. We'll need to change that at some point

Reply:
that's correct. I was assuming that we decided this because we thought if a node is already producing a block or chunk, its low online ratio should be reflected into the BP/CP ratio anyways. But that may not be a true assumption. we can take a look at the data and see if we need to change that.